spec/models/page_spec.rb

```diff
@@ -0,0 +1,105 @@
+require "spec_helper"
+
+RSpec.describe Page, type: :model do
+  describe "unique visitors for phrase" do
+    context "with empty database" do
+      it "returns an empty array when no data exists" do
+        phrase = FactoryBot.build(:phrase)
+        start_date = Date.new(2020, 3, 10)
+        end_date = Date.new(2020, 3, 12)
+
+        result = Page.unique_visitors_for_phrase(phrase, start_date, end_date)
+
+        expect(result).to be_empty
+      end
+    end
+
+    context "with populated database" do
+      before :each do
+        @phrase = FactoryBot.create(:phrase)
+        @visitor = FactoryBot.create(:visitor)
+        page1 = FactoryBot.create(:page, base_path: "/about-government")
+        page2 = FactoryBot.create(:page, base_path: "/government-departments")
+        page3 = FactoryBot.create(:page, base_path: "/overseas-offices")
+        page4 = FactoryBot.create(:page, base_path: "/crown-dependencies")
+
+        create_surveys_for_page(page1, "2020-03-20", 3)
+        create_surveys_for_page(page1, "2020-03-21", 1)
+        create_surveys_for_page(page2, "2020-03-21", 5)
+        create_surveys_for_page(page3, "2020-03-22", 8)
+        create_surveys_for_page(page4, "2020-03-22", 2)
+        
+        phrase2 = FactoryBot.create(:phrase)
+        create_surveys_for_page(page4, "2020-03-20", 6, phrase: phrase2)
+      end
+
+      it "returns an empty array when date range matches no data" do
+        start_date = Date.new(2020, 3, 10)
+        end_date = Date.new(2020, 3, 12)
+
+        result = Page.unique_visitors_for_phrase(@phrase, start_date, end_date)
+
+        expect(result).to be_empty
+      end
+
+      it "returns page data for phrase when date range matches data" do
+        start_date = Date.new(2020, 3, 18)
+        end_date = Date.new(2020, 3, 21)
+
+        result = Page.unique_visitors_for_phrase(@phrase, start_date, end_date)
+
+        expect(result).to eq([["/government-departments", 5], ["/about-government", 4]])
+      end
+
+      it "returns ordered page data for phrase by key and default direction when key is specified and direction is not" do
+        additional_page = FactoryBot.create(:page, base_path: "/what-government-does")
+        create_surveys_for_page(additional_page, "2020-03-20", 2)
+
+        start_date = Date.new(2020, 3, 18)
+        end_date = Date.new(2020, 3, 21)
+
+        result = Page.unique_visitors_for_phrase(@phrase, start_date, end_date, sort_key: "base_path")
+
+        expect(result).to eq([["/what-government-does", 2], ["/government-departments", 5], ["/about-government", 4]])
+      end
+
+      it "returns ordered page data for phrase by direction and default key when direction is specified and key is not" do
+        additional_page = FactoryBot.create(:page, base_path: "/find-help")
```

Comment:
this test and the one prior both have `additional_page`, if there is no need for them to have different base paths could they be defined as a factory and the code re-used?

Reply:
It already is defined as a factory `FactoryBot.define(:page)`. The reason they have different base paths here is because we're testing the ordering.